In [ ]:
Toronto K-Means Clustering assignment

Question 1

In [117]:
import pandas as pd
import requests

In [118]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_url = requests.get(url)
wiki_table = pd.read_html(wiki_url.text)
wiki_table

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [119]:
#getting rid of other tables and keeping only the first
wiki_table = wiki_table[0]
wiki_table

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [120]:
#cleaning data by removing boroughs that are not assigned
wiki_df = wiki_table[wiki_table['Borough'] != 'Not assigned']
wiki_df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [121]:
#grouping by postal code ie M3A
wiki_df = wiki_df.groupby(['Postal Code']).head()
wiki_df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [122]:
wiki_df.Neighbourhood.str.count('Not assigned').sum()

0

In [123]:
wiki_df = wiki_df.reset_index()
wiki_df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [124]:
wiki_df.drop(['index'], axis = 'columns', inplace = True)
wiki_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [125]:
rows = wiki_df.shape

In [126]:
print ('The number of rows in the data frame is', rows)

The number of rows in the data frame is (103, 3)


Question 2

In [127]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [128]:
#lat and long data
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [129]:
print("The shape of our wiki data is: ", wiki_df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [130]:
wiki_df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [131]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [132]:
#adding lat and long to dataset
join_data = wiki_df.join(data.set_index('Postal Code'), on = 'Postal Code', how='inner')
join_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [133]:
join_data.shape

(103, 5)

Question 3

I will apply similar analysis to Toronto as we did to NYC. This includes a K-means clustering and utilising the Foursquare API

In [134]:
!pip install geopy
!pip install folium

In [136]:
#Using Geocoders as opposed to Google API
from geopy.geocoders import Nominatim

In [137]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [138]:
import folium

In [139]:
# Creating the map of Toronto with different neighbourhoods
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(join_data['Latitude'], join_data['Longitude'], join_data['Borough'], join_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto


In [140]:
#Connecting to the Fourquare API

CLIENT_ID = 'LJRIOPWXRMA5UQFWDFWIBDVT3LVFX2V3HIZAYGTHRQRN2O2H' 
CLIENT_SECRET = 'WEJBTAHLZOOPSXTAKADWTTD4D12DSFI0U2NKE2DCOTKFNEW1'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: LJRIOPWXRMA5UQFWDFWIBDVT3LVFX2V3HIZAYGTHRQRN2O2H
CLIENT_SECRET:WEJBTAHLZOOPSXTAKADWTTD4D12DSFI0U2NKE2DCOTKFNEW1


In [141]:
#Testing the df
join_data.loc[0, 'Neighbourhood']

'Parkwoods'

In [142]:
neighbourhood_latitude = join_data.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = join_data.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = join_data.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


In [143]:
#Similar to what we did with NYC - we can create a function to return all venue categories in Toronto

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [144]:
toronto_venues = getNearbyVenues(join_data['Neighbourhood'], join_data['Latitude'], join_data['Longitude'])


Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [145]:
print(toronto_venues.shape)
toronto_venues.head()

(1335, 5)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                               Venue              Venue Category  
0                    Brookbanks Park                        Park  
1                    Brookbanks Pool                        Pool  
2                      Variety Store           Food & Drink Shop  
3  Corrosion Service Company Limited  Construction & Landscaping  
4             Victoria Village Arena                Hockey Arena

In [146]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                               10   
Bathurst Manor, Wilson Heights, Downsview North                      22   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    23   
...                                                                 ...   
Willowdale, Willowdale East                                          30   
Willowdale, Willowdale West                                           5   
Woburn                                                                5   
Woodbine Heights                                                      5   
York Mills West                                                       3   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              5   
Alderwood, Long Branch                                                10   
Bathurst Manor, Wilson Heights, Downsview North                       22   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     23   
...                                                                  ...   
Willowdale, Willowdale East                                           30   
Willowdale, Willowdale West                                            5   
Woburn                                                                 5   
Woodbine Heights                                                       5   
York Mills West                                                        3   

                                                 Venue  Venue Category  
Neighbourhood                                                           
Agincourt                                            5               5  
Alderwood, Long Branch                              10              10  
Bathurst Manor, Wilson Heights, Downsview North     22              22  
Bayview Village                                      4               4  
Bedford Park, Lawrence Manor East                   23              23  
...                                                ...             ...  
Willowdale, Willowdale East                         30              30  
Willowdale, Willowdale West                          5               5  
Woburn                                               5               5  
Woodbine Heights                                     5               5  
York Mills West                                      3               3  

[96 rows x 4 columns]

In [147]:
toronto_venues.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[237 rows x 4 columns]

From this analysis we have learnt that there are 237 different types of venue categories, these range from wing joints to yoga studios.

In [ ]:
Now we want to analyze each neighbourhood by using one hot encoding. 

In [148]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3         Parkwoods                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Truck Stop  \
0                    0             0  ...              0           0   
1                    0             0  ...              0           0   
2                    0             0  ...              0           0   
3                    0             0  ...              0           0   
4                    0             0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

In [149]:
toronto_onehot.shape

(1335, 238)

In [150]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
91                      Willowdale, Willowdale East                0.0   
92                      Willowdale, Willowdale West                0.0   
93                                           Woburn                0.0   
94                                 Woodbine Heights                0.0   
95                                  York Mills West                0.0   

    Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0              0.0      0.0                 0.0             0.0   
1              0.0      0.0                 0.0             0.0   
2              0.0      0.0                 0.0             0.0   
3              0.0      0.0                 0.0             0.0   
4              0.0      0.0                 0.0             0.0   
..             ...      ...                 ...             ...   
91             0.0      0.0                 0.0             0.0   
92             0.0      0.0                 0.0             0.0   
93             0.0      0.0                 0.0             0.0   
94             0.0      0.0                 0.0             0.0   
95             0.0      0.0                 0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.043478           0.0  ...   
..              ...               ...                  ...           ...  ...   
91              0.0               0.0             0.000000           0.0  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   
94              0.0               0.0             0.000000           0.0  ...   
95              0.0               0.0             0.000000           0.0  ...   

    Train Station  Truck Stop  Vegetarian / Vegan Restaurant  \
0             0.0         0.0                            0.0   
1             0.0         0.0                            0.0   
2             0.0         0.0                            0.0   
3             0.0         0.0                            0.0   
4             0.0         0.0                            0.0   
..            ...         ...                            ...   
91            0.0         0.0                            0.0   
92            0.0         0.0                            0.0   
93            0.0         0.0                            0.0   
94            0.0         0.0                            0.0   
95            0.0         0.0                            0.0   

    Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                0.0               0.000000              0.0       0.0   
1                0.0               0.000000              0.0       0.0   
2                0.0               0.000000              0.0       0.0   
3                0.0               0.000000              0.0       0.0   
4                0.0               0.000000              0.0       0.0   
..               ...                    ...              ...       ...   
91               0.0               0.033333              0.0       0.0   
92        

In [151]:
toronto_grouped.shape

(96, 238)

We now have a df size of 96 different neighbourhoods as we just grouped them together.

Now I will print out the top 5 venues in each nighbourhood and put them into a new df

In [153]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')


----Agincourt----
                       venue  freq
0                     Lounge   0.2
1             Breakfast Spot   0.2
2  Latin American Restaurant   0.2
3             Clothing Store   0.2
4               Skating Rink   0.2


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place   0.2
1                 Gym   0.1
2  Athletics & Sports   0.1
3         Coffee Shop   0.1
4                 Pub   0.1


----Bathurst Manor, Wilson Heights, Downsview North----
         venue  freq
0         Bank  0.09
1  Coffee Shop  0.09
2    Pet Store  0.05
3  Gas Station  0.05
4  Bridal Shop  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1   Chinese Restaurant  0.25
2                 Bank  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.09
1  Italian Restaurant  0.09
2         Coffee Shop  0.09
3            Pharmacy  0.04
4  

But now we want that in a df

In [154]:
import numpy as np 

In [155]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [156]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East         Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Clothing Store        Breakfast Spot   
1    Athletics & Sports          Dance Studio           Coffee Shop   
2                  Bank             Pet Store           Supermarket   
3    Chinese Restaurant                  Bank   Japanese Restaurant   
4           Coffee Shop        Sandwich Place           Pizza Place   

  5th Most Common Venue  
0          Skating Rink  
1              Pharmacy  
2      Sushi Restaurant  
3          Dance Studio  
4            Restaurant

Ok, so now we have a dataframe with each nighbourhood and its 5 most common venue categories. The next step will be to cluster these neighbourhoods.

In [157]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [158]:

# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [159]:
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [160]:
toronto_merged = join_data

toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head(3)

Postal Code           Borough              Neighbourhood   Latitude  \
0         M3A        North York                  Parkwoods  43.753259   
1         M4A        North York           Victoria Village  43.725882   
2         M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             1.0                  Park                  Pool   
1 -79.315572             1.0           Pizza Place           Coffee Shop   
2 -79.360636             1.0           Coffee Shop                  Park   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0  Construction & Landscaping     Food & Drink Shop           Yoga Studio  
1       Portuguese Restaurant          Hockey Arena    Dim Sum Restaurant  
2                      Bakery        Breakfast Spot                  Café

In [161]:
#drop NAs to keep data clean
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Now lets plot on a map

In [162]:
#import the packages
import matplotlib.cm as cm
import matplotlib.colors as colors

In [163]:

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=3,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

We now have a map of neighbourhood clusters in Toronto by means of K-means clustering.